### PR TITLE
serve admin-ui directly instead of

### PR DESCRIPTION
**Pull request by bogensberger**
using the elasticsearch site plugin mechanism

### CODE REVIEW

File: admin-ui/src/main/java/io/crate/rest/action/admin/AdminUIFrontpageAction.java [hunk 1]
```diff
@@ -65,13 +79,13 @@ protected void handleRequest(RestRequest request, RestChannel channel, Client cl
...
 
-        if (request.header("accept") != null &&  request.header("accept").toString().contains("application/json")){
+        if (request.header("accept") != null &&  request.header("accept").contains("application/json")){
             crateRestMainAction.handleRequest(request, channel, client);
             return;
         }
 
         BytesRestResponse resp = new BytesRestResponse(RestStatus.TEMPORARY_REDIRECT);
-        resp.addHeader("Location", "/_plugin/crate-admin/");
+        resp.addHeader("Location", "/admin/");
```
Comment (chaudum) [inline admin-ui/src/main/java/io/crate/rest/action/admin/AdminUIFrontpageAction.java]: `/admin/` is used quite often in this class. maybe make it a static string?

File: admin-ui/src/main/java/io/crate/rest/action/admin/AdminUIFrontpageAction.java [hunk 2]
```diff
@@ -24,38 +24,53 @@
...
  */
 public class AdminUIFrontpageAction extends BaseRestHandler {
 
+    private static final String ADMIN_ENPOINT = "/admin/";
     private static final Pattern USER_AGENT_BROWSER_PATTERN = Pattern.compile("(Mozilla|Chrome|Safari|Opera|Android|AppleWebKit)+?[/\\s][\\d.]+");
     private final CrateRestMainAction crateRestMainAction;
     private final RestController controller;
+    private final Environment environment;
+
+    private final AdminUIRequestFilter requestFilter = new AdminUIRequestFilter();
```
Comment (seut) [inline admin-ui/src/main/java/io/crate/rest/action/admin/AdminUIFrontpageAction.java]: should be static

File: admin-ui/src/main/java/io/crate/rest/action/admin/AdminUIFrontpageAction.java [hunk 3]
```diff
@@ -84,4 +99,106 @@ private boolean isBrowser(String headerValue) {
 
         return matcher.matches();
     }
+
+    private class AdminUIRequestFilter extends RestFilter {
+
+        @Override
+        public void process(RestRequest request, RestChannel channel, RestFilterChain filterChain) throws IOException {
+            if(request.rawPath().startsWith(ADMIN_ENPOINT)) {
```
Comment (seut) [inline admin-ui/src/main/java/io/crate/rest/action/admin/AdminUIFrontpageAction.java]: add space after `if`

File: admin-ui/src/main/java/io/crate/rest/action/admin/AdminUIFrontpageAction.java [hunk 4]
```diff
@@ -84,4 +99,106 @@ private boolean isBrowser(String headerValue) {
...
+                serveSite(request, channel);
+            } else {
+                filterChain.continueProcessing(request, channel);
+            }
+        }
+    }
+
+    private void serveSite(RestRequest request, RestChannel channel) throws IOException {
+        if (request.method() != RestRequest.Method.GET) {
```
Comment (seut) [inline admin-ui/src/main/java/io/crate/rest/action/admin/AdminUIFrontpageAction.java]: would be faster to check for GET first and then for the admin endpoint imho ( so I'd move this check to AdminUIRequestFilter.process())

File: admin-ui/src/main/java/io/crate/rest/action/admin/AdminUIFrontpageAction.java [hunk 5]
```diff
@@ -84,4 +99,106 @@ private boolean isBrowser(String headerValue) {
...
+
+        String sitePath = request.rawPath().substring(ADMIN_ENPOINT.length());
+
+        // we default to index.html, or what the plugin provides (as a unix-style path)
+        // this is a relative path under _site configured by the plugin.
+        if (sitePath.length() == 0) {
+            sitePath = "index.html";
+        } else {
+            // remove extraneous leading slashes, its not an absolute path.
+            while (sitePath.length() > 0 && sitePath.charAt(0) == '/') {
```
Comment (seut) [inline admin-ui/src/main/java/io/crate/rest/action/admin/AdminUIFrontpageAction.java]: use StringUtils.stripFront(), does exactly the same ;)

File: admin-ui/src/main/java/io/crate/rest/action/admin/AdminUIFrontpageAction.java [hunk 6]
```diff
@@ -84,4 +99,106 @@ private boolean isBrowser(String headerValue) {
...
+
+        // return not found instead of forbidden to prevent malicious requests to find out if files exist or dont exist
+        if (!Files.exists(file) || FileSystemUtils.isHidden(file) || !file.toAbsolutePath().normalize().startsWith(siteFile.toAbsolutePath().normalize())) {
+            channel.sendResponse(new BytesRestResponse(NOT_FOUND));
+            return;
+        }
+
+        BasicFileAttributes attributes = Files.readAttributes(file, BasicFileAttributes.class);
+        if (!attributes.isRegularFile() && !attributes.isDirectory()) {
+            // If it's not a dir, we send a 403
```
Comment (seut) [inline admin-ui/src/main/java/io/crate/rest/action/admin/AdminUIFrontpageAction.java]: why should we allow directory access? afaik this is not needed and any directory access should result in a 403

File: admin-ui/src/main/java/io/crate/rest/action/admin/AdminUIFrontpageAction.java [hunk 7]
```diff
@@ -84,4 +99,106 @@ private boolean isBrowser(String headerValue) {
...
+
+        try {
+            byte[] data = Files.readAllBytes(file);
+            channel.sendResponse(new BytesRestResponse(OK, guessMimeType(file.toAbsolutePath().toString()), data));
+        } catch (IOException e) {
+            channel.sendResponse(new BytesRestResponse(INTERNAL_SERVER_ERROR));
+        }
+    }
+
+    private String guessMimeType(String path) {
```
Comment (seut) [inline admin-ui/src/main/java/io/crate/rest/action/admin/AdminUIFrontpageAction.java]: could be static

File: admin-ui/src/main/java/io/crate/rest/action/admin/AdminUIFrontpageAction.java [hunk 8]
```diff
@@ -84,4 +99,106 @@ private boolean isBrowser(String headerValue) {
...
+            return "";
+        }
+        return mimeType;
+    }
+
+    private static final Map<String, String> DEFAULT_MIME_TYPES;
+
+    static {
+        // This is not an exhaustive list, just the most common types. Call registerMimeType() to add more.
+        Map<String, String> mimeTypes = new HashMap<>();
```
Comment (seut) [inline admin-ui/src/main/java/io/crate/rest/action/admin/AdminUIFrontpageAction.java]: would prefer using a MapBuilder using the immutableMap method

**Comment by mfussenegger** [comment]
Did you check if the used functionality is still available in ES 5?

**Comment by bogensberger** [comment]
@mfussenegger yes, the functionality i use is still available in ES 5.0

**Pull request merged**